[c10d][nvshmem] modify is_nvshmem_available runtime check to work with static-linked library (#159558)

jushg opened the pull request:
Summary:

Currently this function rely on the logic that we load `libnvshmem_device.a` statically and load `libnvshmem_host.so` at runtime. For loading `libnvshmem.a` (the combine 2 thing together) statically this will fail. Add a section to check if the symbol from host API exist at runtime to check if nvshmem is loaded statically

Test Plan:
CI + sample run

Rollback Plan:

Differential Revision: D79177525




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

Skylion007 commented on torch/csrc/distributed/c10d/symm_mem/nvshmem_extension.cu:
```diff
@@ -28,12 +28,20 @@ namespace c10d::nvshmem_extension {
 
 constexpr int MiB = 1024 * 1024;
 
+extern "C" void nvshmem_init() __attribute__((weak));
```
Wait why do we load the first statically? Seems like a bug that just increases binary size of the wheel?

kwen2501 replied:
@Skylion007 `libnvshmem_device.a` needs to be statically linked AFAIK -- it is a device-side lib that mainly stores NVSHMEM's device-side global states, and there is no .so alternative. It is relatively lightweight tho -- 6.8 MB.

jushg committed:
[c10d][nvshmem] modify is_nvshmem_available runtime check to work with static-linked library (#159561)

Summary:

Currently this function rely on the logic that we load `libnvshmem_device.a` statically and load `libnvshmem_host.so` at runtime. For loading `libnvshmem.a` (the combine 2 thing together) statically this will fail. Add a section to check if the symbol from host API exist at runtime to check if nvshmem is loaded statically

Test Plan:
CI + sample run

Rollback Plan:

Reviewed By: kwen2501

Differential Revision: D79177525
